src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs

```diff
@@ -115,23 +115,25 @@ public static bool IsLockFileStillValid(DependencyGraphSpec dgSpec, PackagesLock
                             // The package spec not found in the dg spec. This could mean that the project does not exist anymore.
                             if (p2pSpec != null)
                             {
-                                var p2pSpecTarget = NuGetFrameworkUtility.GetNearest(p2pSpec.TargetFrameworks, framework.FrameworkName, e => e.FrameworkName);
+                                // This does not consider ATF.
+                                var p2pSpecTargetFrameworkInformation = NuGetFrameworkUtility.GetNearest(p2pSpec.TargetFrameworks, framework.FrameworkName, e => e.FrameworkName);
 
                                 // No compatible framework found
-                                if (p2pSpecTarget != null)
+                                if (p2pSpecTargetFrameworkInformation != null)
                                 {
-                                    var p2pSpecProjectRefTarget = p2pSpec.RestoreMetadata.TargetFrameworks.FirstOrDefault(
-                                        t => EqualityUtility.EqualsWithNullCheck(p2pSpecTarget.FrameworkName, t.FrameworkName));
+                                    // We need to compare the main framework only. Ignoring fallbacks.
+                                    var p2pSpecProjectRestoreMetadataFrameworkInfo = p2pSpec.RestoreMetadata.TargetFrameworks.FirstOrDefault(
+                                        t => NuGetFramework.Comparer.Equals(p2pSpecTargetFrameworkInformation.FrameworkName, t.FrameworkName));
 
-                                    if (p2pSpecProjectRefTarget != null) // This should never happen.
+                                    if (p2pSpecProjectRestoreMetadataFrameworkInfo != null) // This should never happen.
```

Comment:
I think the "This should never happen" comment is in the wrong place. Maybe it belongs in the else block?

Reply:
Yeah, that's a great point. Fixed.